Description: checks whether the C compiler supports the error attribute
  without warning, and disables -Werror if that is not the case as otherwise
  the build will fail.
Note: A better patch is available from the link below.
See: http://lists.gnu.org/archive/html/grub-devel/2010-01/msg00036.html

--- acinclude.m4.orig	2010-01-02 14:42:11.000000000 +0100
+++ acinclude.m4
@@ -458,3 +458,16 @@ else
   AC_MSG_RESULT([no])
 [fi]
 ])
+
+dnl Check if the C compiler supports the error attribute without warning.
+AC_DEFUN(grub_CHECK_ERROR_ATTRIBUTE_WITHOUT_WARNING,[
+AC_MSG_CHECKING([whether `$CC' accepts the error attribute without warning])
+AC_LANG_CONFTEST([[void foo (void) __attribute__ ((error ("foo")));]])
+if AC_TRY_COMMAND([${CC-cc} ${CFLAGS} -S -Werror conftest.c]) && test -s conftest.s; then
+  error_attribute_without_warning=yes
+else
+  error_attribute_without_warning=no
+fi
+rm -f conftest.s
+AC_MSG_RESULT([${error_attribute_without_warning}])
+])
--- configure.ac.orig	2010-01-02 14:42:26.000000000 +0100
+++ configure.ac
@@ -408,10 +408,14 @@ if test x"$sap_possible" = xyes; then
   TARGET_CFLAGS="$TARGET_CFLAGS -mno-stack-arg-probe"
 fi
 
+# Error attribute (used in include/grub/list.h) without warning.
+grub_CHECK_ERROR_ATTRIBUTE_WITHOUT_WARNING
+
 AC_ARG_ENABLE([werror],
 	      [AS_HELP_STRING([--disable-werror],
                              [do not use -Werror when building GRUB])])
-if test x"$enable_werror" != xno ; then
+if test x"$enable_werror" != xno && \
+   test x"$error_attribute_without_warning" = xyes ; then
   TARGET_CFLAGS="$TARGET_CFLAGS -Werror"
 fi
 
